### ISSUE TITLE
File Downloads Should Use the Result<...> type

### ISSUE THREAD

**Issue by RndmTsk** [amosavian/FileProvider#9]
My suggestion is to update downloads (or any response that contains an optional result and error) to take advantage of the Result<...> convention.

Here's a quick overview of how it works:
https://realm.io/news/tryswift-rob-napier-swift-legacy-functional-programming/
(important bit starts around the 12 minute mark, but the entire talk is a good one)

Alamofire uses this paradigm:
https://github.com/Alamofire/Alamofire

**Comment by amosavian** [comment]
@RndmTsk 

This `Either<T>` paradigm can be useful when there is something to return. Here we have only completion handlers because everything is async, even Local provider. When it comes to `Error`, the standard way is using `throw`.

This paradigm is usable in SMB2 digesting methods, which are internal and users will take no advantage.

By the way if you think there is somewhere which it's useful particularly please tell me.

**Comment by RndmTsk** [comment]
I was thinking it could be used in the WebDAVFileProvider's contents(path: completion:)

Cheers,
  Terry Latanville

**Comment by amosavian** [comment]
imo, there is no reason to squeeze function parameters into one.

besides, there may be a situation where both contents and error are nil or both of them are non-nil. There is not a strictly either this or that situation. (a scenario is connection failing in the midst of transferring, where data is trimmed but not necessarily null, and there is an error. a user may use that data and try to initiate another connection from that offset.)

In short, using `Either<T>` system for error handling backs to Swift 1, when there was no `throw` system and using traditional inout error variables was ugly. anyway, coding preferences doesn't mean that something is better than another, maybe someday some changes in api makes using `Either` system better.

**Comment by RndmTsk** [comment]
I didn't realize you would get partial data and still get a callback.

Cheers,
  Terry Latanville
